Automated documentation generation

saher-iq opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/Leenkh3/Firefly/34)
<!-- Reviewable:end -->

Commit:
for testing

saher-iq committed:
Add files via upload

saher-iq committed:
Update Doxyfile

saher-iq committed:
Create doxygen.yml

saher-iq committed:
Update Doxyfile

saher-iq committed:
Update doxygen.yml

saher-iq committed:
Update doxygen.yml

saher-iq committed:
Update doxygen.yml

saher-iq committed:
Update CMakeLists.txt

saher-iq committed:
Update doxygen.yml

jbakosi reviewed (changes requested):
Good work @saher-iq!

* Let's merge in the latest branch `main` to bring this up-to-date with the recent work of rest of the team.
* Let's merge in our existing github CI workflow which already automates builds and testing and make building the documentation target part of it as a new CI matrix entry.

Keep up the good work!

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Leenkh3/Firefly/34#-:-OMk8ekq9ea3EyiKysho:b-r3lyso)* status: all files reviewed, 2 unresolved discussions (waiting on @saher-iq)

___
*[`Doxyfile` line 45 at r1](https://reviewable.io/reviews/Leenkh3/Firefly/34#-OMk8Ihh9Gsvk1Upt72G:-OMk8Ihh9Gsvk1Upt72H:bqvkp6d) ([raw file](https://github.com/Leenkh3/Firefly/blob/347514b8bb310aa7be1cc617375c5948a396b99f/Doxyfile#L45)):*
> ```
> # The default value is: My Project.
> 
> PROJECT_NAME           = "My Project"
> ```

Let's at least add "Firefly" as our project name for the documentation generated.
___
*[`test.txt` line 1 at r1](https://reviewable.io/reviews/Leenkh3/Firefly/34#-OMk97kt9TE0XEcYXhAO:-OMk97ku5sJ8C7tYuF6q:b-f3ntb1) ([raw file](https://github.com/Leenkh3/Firefly/blob/347514b8bb310aa7be1cc617375c5948a396b99f/test.txt#L1)):*
> ```help
> Hello
> ```

Is this file necessary?


<!-- Sent from Reviewable.io -->

jbakosi commented:
Oh, and let's rename the name of the pull request to something more descriptive of the changeset instead of your name. Thanks.

saher-iq committed:
Update Doxyfile

saher-iq committed:
Delete test.txt

extra file

saher-iq committed:
Update CMakeLists.txt

jbakosi reviewed (changes requested):
Good work in progress, @saher-iq! Please address the comments below.

When I try building on this branch, I get:
```sh
$ rm * .ninja_* -rf
$ cmake -GNinja ..
-- The CXX compiler identification is GNU 12.2.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/lib/ccache/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Doxygen: /usr/bin/doxygen (found version "1.9.4") found components: doxygen dot 
CMake Error at CMakeLists.txt:93 (add_dependencies):
  Cannot add target-level dependencies to non-existent target "all".

  The add_dependencies works for top-level logical targets created by the
  add_executable, add_library, or add_custom_target commands.  If you want to
  add file-level dependencies see the DEPENDS option of the add_custom_target
  and add_custom_command commands.


-- Configuring incomplete, errors occurred!
See also "/home/jbakosi/code/Firefly/build/CMakeFiles/CMakeOutput.log".
```
Please fix it: there is no cmake target 'all`.

Let's generate the documentation under the build directory instead of outside of it. Currently, this goes into the top-level project directory. This is not good practice: all build-generated artifacts should go under the build directory.

Please keep this branch updated with main.

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Leenkh3/Firefly/34#-:-OMpdChg3k6viGnn8i2o:b-4lfc90)* status: all files reviewed, 2 unresolved discussions (waiting on @saher-iq)



<!-- Sent from Reviewable.io -->

saher-iq committed:
Update CMakeLists.txt

Removed the add_dependencies(all generate_docs) line: This is unnecessary, as the default all target is already managed by CMake.

Ensured documentation generation happens in the build directory

saher-iq committed:
Update CMakeLists.txt

saher-iq committed:
Update CMakeLists.txt

Keep your branch updated with main: This ensures that your branch includes the latest changes from the team, and any potential issues with integration are avoided.

CI Workflow Integration: Adding the documentation target to the CI matrix ensures that the documentation generation is automatically tested as part of the CI process.

jbakosi reviewed (changes requested):
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Leenkh3/Firefly/34)* status: all files reviewed, 3 unresolved discussions (waiting on @saher-iq)

___
*[`CMakeLists.txt` line 93 at r2](https://reviewable.io/reviews/Leenkh3/Firefly/34#-ONJzOF3BhxkqDXyIw1e:-ONJzOF3BhxkqDXyIw1f:b-dgu8ge) ([raw file](https://github.com/Leenkh3/Firefly/blob/9a5dbbe005e5f090835f2f0624d8afe8aae6553a/CMakeLists.txt#L93)):*
> ```CMake
> )
> 
> add_dependencies(all generate_docs)
> ```

The target  _all_ does not exist, i.e., undefined upstream of the `add_dependencies` command, so cmake triggers an error. But if you remove the `add_custom_target` call, you remove the target for generating the documentation. This not what we want.


<!-- Sent from Reviewable.io -->

saher-iq committed:
Update CMakeLists.txt

Define a custom target for generating documentation

saher-iq committed:
Update CMakeLists.txt

Updated to fix the error

danield0ma commented on CMakeLists.txt:
```diff
@@ -1,16 +1,16 @@
-# Tell cmake the featureset we require
+# تحديد الحد الأدنى لإصدار CMake المطلوب
```
Could you please update the comments to English? 😊

saher-iq committed:
Update CMakeLists.txt

Change comments to English

saher-iq committed:
Merge branch 'main' into saher

saher-iq committed:
Update CMakeLists.txt

saher-iq committed:
Update CMakeLists.txt

saher-iq committed:
Update CMakeLists.txt

jbakosi commented:
The CI helps you by telling you what the problem is: https://github.com/Leenkh3/Firefly/actions/runs/14653483307/job/41124262238?pr=34#step:4:8 (There is leftover git markings after a merge that cmake does not understand.)

saher-iq committed:
Update CMakeLists.txt

saher-iq commented:
I fixed and it run and I create the documentation for all work , but now the system change the version of cmake and I cant update it. 
 CMake 3.15 or higher is required.  You are running version 2.8.12.2

saher-iq committed:
Update cmakebasedCI.yml

necessary fixes

saher-iq committed:
Update cmakebasedCI.yml

Edit workflow file

saher-iq committed:
Update cmakebasedCI.yml

Update

saher-iq committed:
Update CMakeLists.txt

saher-iq commented:
could you please explain to me what is missing ?
